### PR TITLE
Feature/Implement DISC Function

### PR DESCRIPTION
**Pull request by aflorzy**
Implements nearly all of DISC function.

Calculates correctly when using day count bases Actual/Actual, Actual/360, and Actual/365.

US (NASD) 30/360 and European 30/360 day count bases are not yet implemented. The function seems to be calculated more complexly using these two than when compared to the other three. More specifically, the `basisVal` used is not always an integer like the other three (360 or 365)

Meaningful errors are returned according to [Excel documentation](https://support.microsoft.com/en-us/office/disc-function-71fce9f3-3f05-4acf-a5a3-eac6ef4daa53) when invalid arguments are passed.

Tests are implemented for each valid/invalid case. The cases utilizing the 30/360 day count bases are commented out until their implementation is complete.

I originally tried to throw an error if the invalid bases were used, but I couldn't figure out how to catch it in the test. I defaulted with throwing `error.num`. These errors are also commented out in case that is a better solution than returning `error.num`.

### CODE REVIEW

File: test/financial.js [hunk 1]
```diff
@@ -404,9 +404,50 @@ describe('Financial', () => {
     expect(financial.DDB(100000, 1000000, 6, 1, 1.5)).to.equal(0)
   })
 
-  // TODO: implement
   it('DISC', () => {
```
Comment (nicolashefti) [inline test/financial.js:407]: Can you group assertions in a meaningfull way (see tests for TEXT, SUBSTITUE, VALUE) with a first level of `describe` ?

File: src/financial.js [hunk 2]
```diff
@@ -475,24 +475,83 @@ export function DDB(cost, salvage, life, period, factor) {
...
- * @param {*} redemption The security's redemption value per $100 face value.
- * @param {*} basis Optional. The type of day count basis to use.
- * @returns
- */
-export function DISC() {
-  throw new Error('DISC is not implemented')
-}
+  /**
+   * Returns the discount rate for a security.
+   *
```
Comment (nicolashefti) [inline src/financial.js:480]: Maybe warn here about limitated implementation ?

File: src/financial.js [hunk 3]
```diff
@@ -475,24 +475,83 @@ export function DDB(cost, salvage, life, period, factor) {
...
+      return error.num
+    } else if (basis === 1) {
+      basisVal = 365
+    } else if (basis === 2) {
+      basisVal = 360
+    } else if (basis === 3) {
+      basisVal = 365
+    } else if (basis === 4) {
+      // Error: Basis value 4 is not implemented
+      // throw new Error('European 30/360 day count convention is not yet implemented')
```
Comment (nicolashefti) [inline src/financial.js:542]: Could be tested with Chai: https://www.chaijs.com/api/bdd/#method_throw

File: src/financial.js [hunk 4]
```diff
@@ -475,24 +475,83 @@ export function DDB(cost, salvage, life, period, factor) {
...
+      return error.num
+    } else {
+      // Unsupported basis value
+      return error.num
```
Comment (nicolashefti) [inline src/financial.js:546]: Can you remove the lines 520 > 522?

File: src/financial.js [hunk 5]
```diff
@@ -475,24 +475,83 @@ export function DDB(cost, salvage, life, period, factor) {
...
+    redemption = utils.parseNumber(redemption)
+    basis = utils.parseNumber(basis)
+
+    basis = basis || 0;
+
+    if (utils.anyIsError(settlement, maturity, pr, redemption, basis)) {
+      return error.value
+    }
+
+    // Return error if price is less than or equal to zero
```
Comment (nicolashefti) [inline src/financial.js:509]: This comment does not bring much given line 510 is easy to understand.

File: src/financial.js [hunk 6]
```diff
@@ -475,24 +475,83 @@ export function DDB(cost, salvage, life, period, factor) {
...
+    if (pr <= 0) {
+      return error.num
+    }
+
+    // Return error if redemption is less than or equal to zero
```
Comment (nicolashefti) [inline src/financial.js:514]: Same here

**Review by nicolashefti — CHANGES_REQUESTED** [review]
Thanks @aflorzy for the PR. Good work. There are small details to correct. You can add more meaning to the test cases for later extension of the function.

File: src/financial.js (hunk 2 above)
Comment (aflorzy) [reply src/financial.js:482]: Added a simple warning. Can change the format if a different way is preferred

File: test/financial.js (hunk 1 above)
Comment (aflorzy) [reply test/financial.js]: Changed the format of the tests and removed some that were redundant

File: src/financial.js (hunk 3 above)
Comment (aflorzy) [reply src/financial.js]: Got this to work now. Thank you for the link

**Comment by aflorzy** [comment]
Function should be fully implemented at this point. Found that date subtraction uses DAYS360 for 30/360 day count conventions instead of DATEDIF so I was able to get the function to calculate properly using basis values 0 and 4.

**Review by nicolashefti — CHANGES_REQUESTED** [review]
Nice refactoring of the function and test cases. Some minor details to be addressed. Thanks!

**Comment by aflorzy** [comment]
Thanks for reviewing! I have made the requested changes. Same story with #137

**Review by nicolashefti — APPROVED** [review]
Perfect ! Thank you.

**Pull request merged**